File: packages/go-common/logging/killkrill.go
package logging

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"math"
	"net/http"
	"sync"
	"time"
)

const (
	defaultBatchSize     = 100
	defaultFlushInterval = 5 * time.Second
	defaultTimeout       = 10 * time.Second
	defaultMaxRetries    = 3
	eventsPath           = "/api/v1/events"
)

// KillKrillConfig holds configuration for the KillKrill log sink.
type KillKrillConfig struct {
	// Endpoint is the base URL of the KillKrill service (e.g. "https://logs.example.com").
	Endpoint string
	// APIKey is used in the Authorization: Bearer header.
	APIKey string
	// BatchSize is the maximum number of events to send in a single flush. Defaults to 100.
	BatchSize int
	// FlushInterval controls how often the background goroutine flushes the buffer. Defaults to 5s.
	FlushInterval time.Duration
	// UseGRPC is reserved for future gRPC transport support; currently unused.
	UseGRPC bool
	// Timeout is the HTTP client timeout per request. Defaults to 10s.
	Timeout time.Duration
	// MaxRetries is the number of retry attempts on transient failure. Defaults to 3.
	MaxRetries int
}

func (c *KillKrillConfig) applyDefaults() {
	if c.BatchSize <= 0 {
		c.BatchSize = defaultBatchSize
	}
	if c.FlushInterval <= 0 {
		c.FlushInterval = defaultFlushInterval
	}
	if c.Timeout <= 0 {
		c.Timeout = defaultTimeout
	}
	if c.MaxRetries <= 0 {
		c.MaxRetries = defaultMaxRetries
	}
}

// KillKrillSink buffers log events and periodically flushes them to the
// KillKrill ingestion endpoint via HTTP POST with retry and exponential backoff.
type KillKrillSink struct {
	cfg    KillKrillConfig
	client *http.Client

	mu     sync.Mutex
	buffer []map[string]interface{}

	stopCh chan struct{}
	wg     sync.WaitGroup
}

// NewKillKrillSink creates a KillKrillSink and starts a background flush goroutine.
// Call Close() to stop the goroutine and flush remaining events.
func NewKillKrillSink(cfg KillKrillConfig) *KillKrillSink {
	cfg.applyDefaults()

	s := &KillKrillSink{
		cfg:    cfg,
		client: &http.Client{Timeout: cfg.Timeout},
		buffer: make([]map[string]interface{}, 0, cfg.BatchSize),
		stopCh: make(chan struct{}),
	}

	s.wg.Add(1)
	go s.flushLoop()

	return s
}

// Write appends the event to the internal buffer, flushing immediately if the batch is full.
func (s *KillKrillSink) Write(event map[string]interface{}) error {
	s.mu.Lock()
	s.buffer = append(s.buffer, event)
	full := len(s.buffer) >= s.cfg.BatchSize
	s.mu.Unlock()

	if full {
		return s.Flush()
	}
	return nil
}

// Flush drains the buffer and sends all pending events to KillKrill.
func (s *KillKrillSink) Flush() error {
	s.mu.Lock()
	if len(s.buffer) == 0 {
		s.mu.Unlock()
		return nil
	}
	batch := s.buffer
	s.buffer = make([]map[string]interface{}, 0, s.cfg.BatchSize)
	s.mu.Unlock()

	return s.sendWithRetry(batch)
}

// Close stops the background goroutine and flushes any remaining events.
func (s *KillKrillSink) Close() error {
	close(s.stopCh)
	s.wg.Wait()
	return s.Flush()
}

func (s *KillKrillSink) flushLoop() {
	defer s.wg.Done()

	ticker := time.NewTicker(s.cfg.FlushInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			_ = s.Flush()
		case <-s.stopCh:
			return
		}
	}
}

func (s *KillKrillSink) sendWithRetry(batch []map[string]interface{}) error {
	var lastErr error

	for attempt := 0; attempt <= s.cfg.MaxRetries; attempt++ {
		if attempt > 0 {
			backoff := time.Duration(math.Pow(2, float64(attempt-1))) * 100 * time.Millisecond
			time.Sleep(backoff)
		}

		if err := s.send(batch); err != nil {
			lastErr = err
			continue
		}
		return nil
	}

	return fmt.Errorf("killkrill: all %d attempts failed, last error: %w", s.cfg.MaxRetries+1, lastErr)
}

func (s *KillKrillSink) send(batch []map[string]interface{}) error {
	payload, err := json.Marshal(batch)
	if err != nil {
		return fmt.Errorf("killkrill: marshal batch: %w", err)
	}

	url := s.cfg.Endpoint + eventsPath
	req, err := http.NewRequestWithContext(context.Background(), http.MethodPost, url, bytes.NewReader(payload))
	if err != nil {
		return fmt.Errorf("killkrill: build request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+s.cfg.APIKey)

	resp, err := s.client.Do(req)
	if err != nil {
		return fmt.Errorf("killkrill: http request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("killkrill: unexpected status %d", resp.StatusCode)
	}

	return nil
}

File: packages/python-utils/src/penguintechinc_utils/killkrill.py
"""
KillKrill sink — ships structured log events to the KillKrill log aggregation service.

Events are buffered in memory and flushed by a background thread either when
the batch fills up or the flush interval elapses. Failed flushes are retried
with exponential backoff up to max_retries times before the batch is dropped.
"""

import json
import logging
import threading
import time
from dataclasses import dataclass, field
from typing import Any

import httpx

logger = logging.getLogger(__name__)


@dataclass(slots=True)
class KillKrillConfig:
    """
    Configuration for the KillKrill log aggregation sink.

    Attributes:
        endpoint: Base URL of the KillKrill service (e.g. "https://logs.example.io").
        api_key: Bearer token used to authenticate with the service.
        batch_size: Maximum number of events per HTTP request (default: 100).
        flush_interval: Seconds between automatic flushes (default: 5.0).
        use_grpc: Reserved for future gRPC transport support (default: False).
        timeout: HTTP request timeout in seconds (default: 10.0).
        max_retries: Maximum delivery attempts per batch before dropping (default: 3).
    """

    endpoint: str
    api_key: str
    batch_size: int = 100
    flush_interval: float = 5.0
    use_grpc: bool = False
    timeout: float = 10.0
    max_retries: int = 3


@dataclass
class _Buffer:
    """Thread-safe event buffer."""

    events: list[dict[str, Any]] = field(default_factory=list)
    lock: threading.Lock = field(default_factory=threading.Lock)


class KillKrillSink:
    """
    Buffers log events and ships them in batches to the KillKrill service.

    A background daemon thread wakes every flush_interval seconds and sends
    any buffered events. The buffer is also flushed eagerly when it reaches
    batch_size. Call close() for a clean shutdown that flushes remaining events.

    Args:
        config: KillKrillConfig describing the remote endpoint and tuning knobs.
    """

    def __init__(self, config: KillKrillConfig) -> None:
        self._config = config
        self._buffer = _Buffer()
        self._client = httpx.Client(
            headers={
                "Authorization": f"Bearer {config.api_key}",
                "Content-Type": "application/json",
            },
            timeout=config.timeout,
        )
        self._stop_event = threading.Event()
        self._flush_thread = threading.Thread(
            target=self._flush_loop,
            name="killkrill-flush",
            daemon=True,
        )
        self._flush_thread.start()

    # ------------------------------------------------------------------
    # Sink Protocol
    # ------------------------------------------------------------------

    def emit(self, event: dict[str, Any]) -> None:
        """Buffer an event, flushing immediately if the batch is full."""
        with self._buffer.lock:
            self._buffer.events.append(event)
            should_flush = len(self._buffer.events) >= self._config.batch_size

        if should_flush:
            self._flush()

    def flush(self) -> None:
        """Flush all buffered events to the remote service."""
        self._flush()

    def close(self) -> None:
        """Stop the background thread and flush remaining events."""
        self._stop_event.set()
        self._flush_thread.join(timeout=self._config.timeout + 1)
        self._flush()
        self._client.close()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _flush_loop(self) -> None:
        """Background thread: flush on interval until stop is signalled."""
        while not self._stop_event.wait(timeout=self._config.flush_interval):
            self._flush()

    def _flush(self) -> None:
        """Drain the buffer and deliver events with retry/backoff."""
        with self._buffer.lock:
            if not self._buffer.events:
                return
            batch = self._buffer.events
            self._buffer.events = []

        self._deliver_with_retry(batch)

    def _deliver_with_retry(self, batch: list[dict[str, Any]]) -> None:
        """Attempt delivery up to max_retries times with exponential backoff."""
        url = f"{self._config.endpoint}/api/v1/events"
        payload = json.dumps(batch)

        for attempt in range(1, self._config.max_retries + 1):
            try:
                response = self._client.post(url, content=payload)
                response.raise_for_status()
                return
            except httpx.HTTPError as exc:
                if attempt == self._config.max_retries:
                    logger.warning(
                        "KillKrillSink: dropping %d events after %d failed attempts: %s",
                        len(batch),
                        attempt,
                        exc,
                    )
                    return

                backoff = 2 ** (attempt - 1)
                logger.debug(
                    "KillKrillSink: attempt %d failed, retrying in %.1fs: %s",
                    attempt,
                    backoff,
                    exc,
                )
                time.sleep(backoff)
